demos/bounded_heat/bounded_heat.py.rst

```diff
@@ -0,0 +1,202 @@
+Solving the Heat Equation with Bounds Constraints
+=================================================
+
+In this demo we solve the simple heat equation with bounds constraints applied uniformly in time and space. 
+Consider the simple heat equation on :math:`\Omega = [0,1]\times [0,1]`, with boundary :math:`\Gamma`:
+
+.. math::
+
+    u_t - \Delta u &= f
+
+    u & = g \quad \textrm{on}\ \Gamma
+
+for some known functions :math:`f` and :math:`g`. The solution will be some function :math:`u\in V`, for 
+a suitable function space :math:`V`.
+
+The weak form is found by multiplying by an arbitrary test function :math:`v\in V` and integrating over :math:`\Omega`. 
+We then have the variational problem of finding :math:`u:[0,T]\rightarrow V` such that 
+.. math::
+
+    (u_t, v) + (\nabla u, \nabla v) = (f, v)
+
+This demo uses particular choices of the functions :math:`f` and :math:`g` to be defined below.
+
+First, we must import firedrake and certain items from Irksome: ::
+
+    from firedrake import *
+    from irksome import (Dt, MeshConstant, RadauIIA, TimeStepper)
+
+We also need some UFL tools in order to manufacture a solution: ::
+
+    from ufl.algorithms import expand_derivatives
+
+Finally, numpy provides us with the upper bound of infinity: ::
+
+    import numpy as np
+
+We first define the mesh and the necessary function space. We choose 
+quadratic Bernstein polynomials to support bounds-constraints in space: ::
+
+    N = 32
+
+    msh = UnitSquareMesh(N, N)
+    V = FunctionSpace(msh, "Bernstein", 2)
+
+In order to enforce bounds constraints in time, we must utilize a collocation method. 
+In this demo, we will time-step using the L-stable, fully implicit, 2-stage RadauIIA 
+Runge-Kutta method. The bounds will be passed as an argument to the 
+:meth:`~.TimeStepper.advance` method. We now define the Butcher Tableau and variables to store the 
+time step, current time, and final time: ::
+
+    butcher_tableau = RadauIIA(2)
```

Comment:
Important note: I think that RadauIIA(1)/backward Euler will guarantee the bounds constraints, but higher-order methods don't.  This is worth putting in the discussion.

Reply:
While the exact solution chosen is positive, the manufactured RHS is not.  Unfortunately, RadauIIA(1) does not provide a positive approximation in this case.  The same initial condition with no forcing would also work (and backward Euler would be positive), but then there would be no exact solution to compare the approximations against.  I can find a more suitable example if we want to highlight the difference between backward Euler/higher-order methods.

Reply:
This might not be wrong -- the order barrier says that higher-order can't preserve bounds, not that backward Euler will for any problem.  I think we're fine here.